persistent/Database/Persist/Quasi/Internal/ModelParser.hs

```diff
@@ -0,0 +1,617 @@
+{-# LANGUAGE DeriveLift #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TupleSections #-}
+
+module Database.Persist.Quasi.Internal.ModelParser
+    ( SourceLoc (..)
+    , Token (..)
+    , tokenContent
+    , anyToken
+    , ParsedEntityDef
+    , parsedEntityDefComments
+    , parsedEntityDefEntityName
+    , parsedEntityDefIsSum
+    , parsedEntityDefEntityAttributes
+    , parsedEntityDefFieldAttributes
+    , parsedEntityDefExtras
+    , parsedEntityDefSpan
+    , parseSource
+    , memberBlockAttrs
+    , ParseResult
+    , CumulativeParseResult
+    , toCumulativeParseResult
+    , cumulativeData
+    , renderErrors
+    , runConfiguredParser
+    , initialExtraState
+    ) where
+
+import Control.Monad (void)
+import Control.Monad.Trans.State
+import Data.Either (partitionEithers)
+import Data.Foldable (fold)
+import Data.List (intercalate)
+import Data.List.NonEmpty (NonEmpty (..))
+import qualified Data.List.NonEmpty as NEL
+import qualified Data.Map as M
+import Data.Maybe
+import Data.Text (Text)
+import qualified Data.Text as Text
+import Data.Void
+import Database.Persist.Types
+import Database.Persist.Types.SourceSpan
+import Language.Haskell.TH.Syntax (Lift)
+import Text.Megaparsec hiding (Token)
+import Text.Megaparsec.Char
+import qualified Text.Megaparsec.Char.Lexer as L
+
+-- We'll augment the parser with extra state to accumulate comments seen during parsing.
+-- Comments are lexed as whitespace, but will be used to generate documentation later.
+data ExtraState = ExtraState
+    { esPositionedCommentTokens :: [(SourcePos, CommentToken)]
+    , esLastDocumentablePosition :: Maybe SourcePos
+    }
+
+-- @since 2.16.1.0
+initialExtraState :: ExtraState
+initialExtraState =
+    ExtraState
+        { esPositionedCommentTokens = []
+        , esLastDocumentablePosition = Nothing
+        }
+
+type Parser a =
+    StateT
+        ExtraState
+        (Parsec Void String)
+        a
+
+type EntityParseError = ParseErrorBundle String Void
+type InternalParseResult a = Either EntityParseError (a, ExtraState)
+
+type ParseResult a = Either EntityParseError a
+type CumulativeParseResult a = Either [EntityParseError] a
+
+cumulativeData :: (Monoid a) => CumulativeParseResult a -> a
+cumulativeData cpr = case cpr of
+    Left _ -> mempty
+    Right r -> r
```

Comment:
You almost never want a function signature like `MaybeOrEither a -> a`. Instead, users should expect to `case` on the `Either` and handle that `Left` themselves (or use a function like `either`, or `for_`, etc to handle it)